library/std/src/gc.rs

```diff
@@ -175,11 +201,76 @@ pub fn thread_registered() -> bool {
     unsafe { bdwgc::GC_thread_is_registered() != 0 }
 }
 
+// Ring buffer of objects to be finalized.
+//
+// This is considered a root for the Gc because the 'to-be-finalised' objects are otherwise
+// unreachable and must be kept alive until their finaliser has run.
+static FIN_Q: Mutex<VecDeque<Gc<()>>> = Mutex::new(VecDeque::new());
```

Comment:
Doesn't this mean we're doing heap allocations *during* GC? I'm also not sure about the performance properties of `VecDeque`, particularly as I guess that we often shove huge numbers of objects in here. Update: ah, I see on line 225 that we ensure we don't enlarge it! What happens then?!

I also guess that we're going to be heavily punished for the `Mutex` here, as presumably parallel GC ends up locking/unlocking this heavily during particular parts of the GC cycle. I don't know if there's a good way of measuring that.

I don't think we should do this in this PR but I think we could fairly easily do this in the next PR:

1. Each GC thread has a `thread_local` queue of objects to finalise. That could be a `VecDeque` but (because of the "what happens if we run out of space in the `VecDeque` problem) I think that might better be an `mmap`ed block (or similar) under our control.
2. At the end of the GC cycle (or when that local queue is full) we flush it to a global "queue of queues", which is something like `Mutex<Vec<FinalisationQueue>>`.

The aim here being that our number of locks goes from O(number_of_objects_to_finalise) to O(number_of_GC_threads).

Reply:
> Update: ah, I see on line 225 that we ensure we don't enlarge it! What happens then?!

We check whether we need to enlarge the queue on each allocation because this is the only time it would be safe to do so. I'll be the first to admit I don't like this, but I couldn't see another a way around accidentally growing the ring buffer while inside the STW pause.

> I also guess that we're going to be heavily punished for the Mutex here, as presumably parallel GC ends up locking/unlocking this heavily during particular parts of the GC cycle. I don't know if there's a good way of measuring that.

Profile data still shows that we spend a lot of time in `futex_wait` syscalls -- especially from within the GC. However, we're punished for this `Mutex` more because of contention between the finaliser thread and user mutator threads. The more _concurrent_ mutator (application) threads we have, the worse this gets. However,  this does not affect parallel collector threads (which do marking) because they don't lock/unlock this mutex (and Rust mutexes don't heap allocate anymore).

I agree your suggested approach is better. It's similar to what I had lined up (but still suffers from the `VecDeque` grow problem). I'm not sure what you mean by this though:

>  I think that might better be an mmaped block (or similar) under our control.

Do you mean `mmap` a block of pointers to objects that need finalising with the system allocator (i.e. outside of Boehm)? We'd still need to grow this block dynamically, and we also need Boehm to know about it so that it can treat it as a source of roots.

> At the end of the GC cycle (or when that local queue is full) we flush it to a global "queue of queues", which is something like Mutex<Vec<FinalisationQueue>>.

Right, so we batch up finalisation in pre-defined chunks that get siphoned off to the finaliser thread once they reach some predetermined size?

Reply:
>We check whether we need to enlarge the queue on each allocation because this is the only time it would be safe to do so.

Right, so this isn't something we ideally want to do in the fast path.

> Right, so we batch up finalisation in pre-defined chunks that get siphoned off to the finaliser thread once they reach some predetermined size?

Yep, that's the key idea!

>Do you mean mmap a block of pointers to objects that need finalising using system allocator (i.e. outside of Boehm)?

Assuming we can teach Boehm about the `mmap`ed block of pointers somehow, that would be my suggestion. But there are other conceivable ways which achieve the same outcome.

> We'd still need to grow this block dynamically, and we also need Boehm to know about it so that it can treat it as a source of roots.

Yes.

Reply:
Ok, with this + https://github.com/softdevteam/alloy/pull/114#discussion_r1544803488 I think we might be able to eliminate synchronisation around the `FIN_Q`. Ok to do this in the next PR?

Reply:
Works for me.